aptos-move/aptos-vm/src/move_vm_ext/session/user_transaction_sessions/prologue.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    move_vm_ext::{
+        session::{
+            respawned_session::RespawnedSession, user_transaction_sessions::user::UserSession,
+        },
+        AptosMoveResolver, SessionId,
+    },
+    transaction_metadata::TransactionMetadata,
+    AptosVM,
+};
+use aptos_vm_types::{change_set::VMChangeSet, storage::change_set_configs::ChangeSetConfigs};
+use derive_more::{Deref, DerefMut};
+use move_core_types::vm_status::VMStatus;
+
+#[derive(Deref, DerefMut)]
+pub struct PrologueSession<'r, 'l> {
+    #[deref]
+    #[deref_mut]
+    session: RespawnedSession<'r, 'l>,
+}
+
+impl<'r, 'l> PrologueSession<'r, 'l> {
+    pub fn new(
+        vm: &'l AptosVM,
+        txn_meta: &'l TransactionMetadata,
+        resolver: &'r impl AptosMoveResolver,
+    ) -> Result<Self, VMStatus> {
+        let session_id = SessionId::prologue_meta(txn_meta);
+        let session = RespawnedSession::spawn(vm, session_id, resolver, VMChangeSet::empty())?;
+
+        Ok(Self { session })
+    }
+
+    pub fn into_user_session(
+        self,
+        vm: &'l AptosVM,
+        txn_meta: &'l TransactionMetadata,
+        resolver: &'r impl AptosMoveResolver,
+        gas_feature_version: u64,
+        change_set_configs: &ChangeSetConfigs,
+    ) -> Result<(VMChangeSet, UserSession<'r, 'l>), VMStatus> {
+        let Self { session } = self;
+
+        if gas_feature_version >= 1 {
+            // Create a new session so that the data cache is flushed.
+            // This is to ensure we correctly charge for loading certain resources, even if they
+            // have been previously cached in the prologue.
+            //
+            // TODO(Gas): Do this in a better way in the future, perhaps without forcing the data cache to be flushed.
+            // By releasing resource group cache, we start with a fresh slate for resource group
+            // cost accounting.
+
+            let change_set = session.finish_with_squashed_change_set(change_set_configs, false)?;
+            resolver.release_resource_group_cache();
+
+            Ok((
+                change_set.clone(),
+                UserSession::new(vm, txn_meta, resolver, change_set)?,
```

Comment:
I thought we can get some perf win by continuing the session when it doesn't abort and only materialize the change set on failure, but I guess that's too hard (I gave up on that)

Reply:
We can't because when the user session aborts you can't distinguish the changes from the prologue and the user txn?